File: app/src/main/java/app/trigger/nuki/NukiReadLockStateCallback.java
package app.trigger.nuki;

import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;

import app.trigger.DoorReply.ReplyCode;
import app.trigger.Log;
import app.trigger.NukiDoorSetup;
import app.trigger.OnTaskCompleted;
import app.trigger.Utils;


class NukiReadLockStateCallback extends NukiCallback {
    private static final String TAG = "ReadLockStateCallback";
    long auth_id;
    byte[] shared_key;
    byte[] data;

    NukiReadLockStateCallback(int setup_id, OnTaskCompleted listener, NukiDoorSetup setup) {
        super(setup_id, listener, KEYTURNER_SERVICE_UUID, KEYTURNER_USDIO_XTERISTIC_UUID);
        this.shared_key = Utils.hexStringToByteArray(setup.shared_key);
        this.auth_id = setup.auth_id;
    }

    public void onConnected(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
        Log.d(TAG, "onConnected");
        NukiCommand.NukiRequest nr = new NukiCommand.NukiRequest(0x0c);
        byte[] request = NukiRequestHandler.encrypt_message(this.shared_key, this.auth_id, nr.generate(), null);
        characteristic.setValue(request);
        boolean ok = gatt.writeCharacteristic(characteristic);
        if (!ok) {
            Log.e(TAG, "initial writeCharacteristic failed");
            closeConnection(gatt);
        }
    }

    @Override
    public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
        Log.d(TAG, "onCharacteristicChanged, uiid: " + characteristic.getUuid() + ": " + Utils.byteArrayToHexString(characteristic.getValue()));
        if (data == null) {
            data = characteristic.getValue();
        } else {
            data = NukiTools.concat(data, characteristic.getValue());
        }

        byte[] message = NukiRequestHandler.decrypt_message(shared_key, data);
        NukiCommand m = NukiRequestHandler.parse(message);

        if (m == null) {
            Log.d(TAG, "NukiCommand is null");
            return;
        } else {
            data = null;
        }

        if (m instanceof NukiCommand.NukiStates) {
            NukiCommand.NukiStates ns = (NukiCommand.NukiStates) m;

            String extra = "";
            if (ns.battery_critical == 0x01) {
                extra = " (Battery Critical!)";
            }

            listener.onTaskResult(
                setup_id, ReplyCode.SUCCESS, NukiTools.getLockState(ns.lock_state) + extra
            );

            // do not wait until the Nuki closes the connection
            closeConnection(gatt);
        } else if (m instanceof NukiCommand.NukiError) {
            NukiCommand.NukiError ns = (NukiCommand.NukiError) m;
            this.listener.onTaskResult(setup_id, ReplyCode.REMOTE_ERROR, ns.asString());
            closeConnection(gatt);
        } else {
            Log.e(TAG, "Unhandled command.");
            closeConnection(gatt);
        }
    }
}

File: app/src/main/kotlin/app/trigger/nuki/NukiReadLockStateCallback.kt
package app.trigger.nuki

import app.trigger.Utils.byteArrayToHexString
import app.trigger.Utils.hexStringToByteArray
import app.trigger.DoorReply.ReplyCode
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGatt
import app.trigger.*

internal class NukiReadLockStateCallback(door_id: Int, listener: OnTaskCompleted, setup: NukiDoor) : NukiCallback(door_id, listener, NukiCallback.Companion.KEYTURNER_SERVICE_UUID, NukiCallback.Companion.KEYTURNER_USDIO_XTERISTIC_UUID) {
    var auth_id: Long
    var shared_key: ByteArray
    var data: ByteArray = ByteArray(0)

    override fun onConnected(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
        Log.d(TAG, "onConnected")
        val nr = NukiCommand.NukiRequest(0x0c)
        val request: ByteArray? = NukiRequestHandler.encrypt_message(shared_key, auth_id, nr.generate(), null)
        characteristic.value = request
        val ok = gatt.writeCharacteristic(characteristic)
        if (!ok) {
            Log.e(TAG, "initial writeCharacteristic failed")
            closeConnection(gatt)
        }
    }

    override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
        Log.d(TAG, "onCharacteristicChanged, uiid: ${characteristic.uuid}: ${byteArrayToHexString(characteristic.value)}")
        data = if (data == null) {
            characteristic.value
        } else {
            NukiTools.concat(data, characteristic.value)
        }
        val message: ByteArray? = NukiRequestHandler.decrypt_message(shared_key, data)
        val m: NukiCommand? = NukiRequestHandler.parse(message)
        if (m == null) {
            Log.d(TAG, "NukiCommand is null")
            return
        } else {
            data = ByteArray(0)
        }

        if (m is NukiCommand.NukiStates) {
            val ns = m
            var extra = ""
            if (ns.battery_critical == 0x01) {
                extra = " (Battery Critical!)"
            }
            listener.onTaskResult(
                    door_id, ReplyCode.SUCCESS, NukiTools.getLockState(ns.lock_state) + extra
            )

            // do not wait until the Nuki closes the connection
            closeConnection(gatt)
        } else if (m is NukiCommand.NukiError) {
            listener.onTaskResult(door_id, ReplyCode.REMOTE_ERROR, m.asString())
            closeConnection(gatt)
        } else {
            Log.e(TAG, "Unhandled command.")
            closeConnection(gatt)
        }
    }

    companion object {
        private const val TAG = "ReadLockStateCallback"
    }

    init {
        shared_key = hexStringToByteArray(setup.shared_key)
        auth_id = setup.auth_id
    }
}